library/core/src/num/diy_float.rs

```diff
@@ -21,61 +21,29 @@ pub struct Fp {
 
 impl Fp {
     /// Returns a correctly rounded product of itself and `other`.
-    pub fn mul(&self, other: &Fp) -> Fp {
-        const MASK: u64 = 0xffffffff;
-        let a = self.f >> 32;
-        let b = self.f & MASK;
-        let c = other.f >> 32;
-        let d = other.f & MASK;
-        let ac = a * c;
-        let bc = b * c;
-        let ad = a * d;
-        let bd = b * d;
-        let tmp = (bd >> 32) + (ad & MASK) + (bc & MASK) + (1 << 31) /* round */;
-        let f = ac + (ad >> 32) + (bc >> 32) + (tmp >> 32);
+    pub fn mul(self, other: Self) -> Self {
+        let (lo, hi) = self.f.widening_mul(other.f);
+        let f = hi + (lo >> 63) /* round */;
         let e = self.e + other.e + 64;
-        Fp { f, e }
+        Self { f, e }
     }
```

Comment:
can you self-review the sets of lines on this PR that will explain why these new changes are correct?

Reply:
If one squints at the old code, one can see a 64x64->128 integer multiplication using the schoolbook algorithm.
The operands are split into two 32-bit chunks each (lines 26-29) and multiplied with each other (lines 30-29) for 64-bit results. Since this is fp-arithmetic, only the high part of the multiplication gets used and the lower part rounded.

Rounding is done by adding the top bit from the lower half to upper half of the product. The old approach does an addition by `1 << 31` which leverages carry propagation to push the bit upwards. The new approach extracts the bit using a shift.

I think `u128` values were not available when the original algorithm was written, which would have made the original approach the best possible at the time.

Reply:
like, I would not have guessed this is `widening_mul` by looking at it and I'm kind of curious what called it out to you.

Reply:
Part of it is because I'm [a bit familiar](https://github.com/rust-lang/rust/pull/136409) with these kinds of multiplications,
I use a similar approch in a private project of mine.

Part of it is noticing that it is floating point multiplication according to the comments. How else would it be implemented?

Reply:
neat.